Add a Prototype for API Documentations

dhenneke opened the pull request:
Here is our first implementation for the API documentation plugin aka api-docs. See [here](https://github.com/spotify/backstage/issues/627#issuecomment-662972470) for the initial discussion by @Fox32.

For now API entities can be added to the catalog and referenced from a component. They are then displayed under the API tab in the details view. In addition, a standalone API list is added. See the screenshots for some details:

<p>
<img src="https://user-images.githubusercontent.com/720821/88403868-64b82180-cdcd-11ea-9f80-78fe1f23fa5e.png" width=400>
<img src="https://user-images.githubusercontent.com/720821/88403886-697cd580-cdcd-11ea-8922-6c4637ca35e4.png" width=400>
</p>
<p>
<img src="https://user-images.githubusercontent.com/720821/88403895-6b469900-cdcd-11ea-8604-2523f65eee4e.png" width=400>
<img src="https://user-images.githubusercontent.com/720821/88403894-6aae0280-cdcd-11ea-9f7b-7fd8b42af3cf.png" width=400>
</p>
<img src="https://user-images.githubusercontent.com/720821/88403892-6a156c00-cdcd-11ea-9ebc-d1df52f620a0.png" width=400>

We ran into some problems that are not solved till now. Probably these are things we can't resolve right now, but should be the next steps:
- We added two external react components to display the documentation. Styling of these components could be improved by introducing own CSS files. Especially dark mode isn't supported right now. The question is, how can we style it using the backstage theme?
- Right now it is difficult to add new plugins to the catalog details. We added it directly, but it feels like quite a hard dependency. In addition the tab navigation isn't ready yet, but [others are struggeling](https://github.com/spotify/backstage/pull/1707) too.
- Adding a API list to the plugin would be nice, but brings some problems. We would have a cycling dependency between the plugin and the catalog-plugin (the scaffolder-plugin has the same problem). Do we need a package that contains central catalog components like the `catalogAPiRef`?
- We added AsyncAPI support, but have trouble running the jest tests with it. It's not clear what exactly the error during test execution means: "Jest encountered an unexpected token"
- An ownership property for the API entity might be interesting.

We know the PR is quite big but we think that it is a good start for discussions on a working prototype. If you are interested, we can split it up a bit more along the commits and contribute the changes in individual PRs.

stefanalund commented:
Rockstars. Thanks a lot! 💯🥳👏🏼

Rugvip commented on plugins/api-docs/src/components/ApiEntityPage/ApiEntityPage.tsx:
```diff
@@ -0,0 +1,131 @@
+  );
+
+  return (
+    <Page theme={getPageTheme(entity)}>
```
This definitely needs a nicer API (in core) imo, but not for this PR 😅

mbruggmann commented:
Can't really comment on the TypeScript implementation, but from a modelling/metadata perspective this looks pretty great! As @Rugvip mentioned, we might want to give the `spec.implementsApis` part another thought but I'd vote to start with this for now and re-evaluate later if we need to. Thanks so much for getting this going!

dhenneke commented on plugins/api-docs/src/components/ApiEntityPage/ApiEntityPage.test.tsx:
```diff
@@ -0,0 +1,100 @@
+ * limitations under the License.
+ */
+
+jest.mock('react-router-dom', () => {
```
I copied it from the catalog plugin 😀. But I can change it.

Rugvip replied:
Dammit, ok :grin:

It's fine to leave as is

dhenneke commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -312,6 +316,14 @@ Apart from being a string, the software catalog leaves the format of this field
 open to implementers to choose. Most commonly, it is set to the ID or email of a
 group of people in an organizational structure.
 
+### `spec.implementsApis` [optional]
```
Yes. Since [ADR 005](https://github.com/spotify/backstage/blob/master/docs/architecture-decisions/adr005-catalog-core-entities.md#decision) also intends the "consumes" relation it might be good to have a more general approach in the future. As soon as some kind of versioning is introduced, this simple modeling will also not suffice. But we will see then.

dhenneke commented:
I fixed the issue with the failing build by updating the jest config (see last commit):
```diff
- '/node_modules/(?!.*\\.(?:esm\\.js|bmp|gif|jpg|jpeg|png|frag|xml|svg)$)',
+ '/node_modules/(?!@kyma-project/asyncapi-react/)(?!.*\\.(?:esm\\.js|bmp|gif|jpg|jpeg|png|frag|xml|svg)$)',
```
The npm package seems to be a bit odd...

I am not sure if this is a smart solution, especially since I changed in the cli package :-/.

Rugvip commented on packages/cli/config/jest.js:
```diff
@@ -48,8 +48,9 @@ async function getConfig() {
+    // The @kyma-project/asyncapi-react library needs to be transformed.
     transformIgnorePatterns: [
-      '/node_modules/(?!.*\\.(?:esm\\.js|bmp|gif|jpg|jpeg|png|frag|xml|svg)$)',
+      '/node_modules/(?!@kyma-project/asyncapi-react/)(?!.*\\.(?:esm\\.js|bmp|gif|jpg|jpeg|png|frag|xml|svg)$)',
```
Yeah this is idd unfortunate. The page is using esm modules but really doesn't signal that in any way as far as I can tell. I don't wanna block this PR because of the issue so fine with override for now, but we do need to figure out a long-term solution. It would be possible to add this to `.jest.transformIgnorePatterns` inside `package.json` of the `api-docs` package, but that would have to be kept in sync with updates to this config so it isn't great either.

dhenneke replied:
@Rugvip I can also try to fix the original package so we don't get an issue here. But tbh, I am not really deep into jest and esm modules. Do you have a clue what might be wrong there? Would you say this issue is specific to the configuration in the backstage repository?

Rugvip replied:
It's a somewhat common override that's needed when using jest. The package is built in a way that makes it tricky to consume, and could definitely be left as feedback to the package author.

A common way to solve the issue is to build both CommonJS and ESModule outputs, and then point to each using `main` and `module` fields in `package.json`. It's puts more requirements on the build tooling though. For example, we do that here: https://github.com/spotify/backstage/blob/c4b6c1397577b89f117a2f663088315dda75aaa7/packages/catalog-model/package.json#L11

Fox32 commented:
Just to let you know: we are still interested, but on vacation right now. We will continue with it next week 😊

Rugvip commented:
@Fox32 thanks you for the update! Many vacations on our side too :grin:

stefanalund commented:
@dhenneke what is the status of this PR? Excited to get it merged 😃

dhenneke commented:
The build should work now. Strangely, the "Frontend CI" workflow has not been triggered so the tests weren't executed 😕. I wonder if it is connected to https://github.com/spotify/backstage/commit/9d1f531a183e56801419d50b6d5e00bc91f0cf34.

Apart from that: Are there any blockers before merging this PR? If our changes in the Entity-tab-behavior are conflicting, we will be happy to remove them in favor of the work done in #1835. We could integrate that separately after that has been merged. Just leave a short note on what is missing on the road of getting this merged.

Rugvip commented:
@dhenneke Will have a proper look in a bit, but iirc we're just waiting for conflicts to be resolved and should then be good to merge 👍 

Workflows are indeed not triggering correctly, fix in it's way 😅 https://github.com/spotify/backstage/pull/1918

Can't comment on the tab implementation, what do you think @hooloovooo ?

dhenneke commented:
@Rugvip did you have time to have a look?

dhenneke commented:
Builds are green 😲

Rugvip commented on plugins/api-docs/README.md:
```diff
@@ -0,0 +1,9 @@
+
+## Links
+
+- (The Backstage homepage)[https://backstage.io]
```
Can fix later, let's :shipit: :grin:

Fox32 commented:
🥳
